### PR TITLE
ARROW-10051: [C++][Compute] Move kernel state when merging

### PR DESCRIPTION
**Pull request by cyb70289**
Aggregate kernel consumes one batch and output a `state`, which is then
merged with other batches. Currently, `state` parameter is defined as
`const KernelState&` in `merge` interface. It may cause unnecessary data
copying for kernels with non-trivial `states`.
E.g., mode kernel maintains a value:count map in its `state` structure,
by removing `const`, we can move the map rather than copy it.

### CODE REVIEW

**Comment by cyb70289** [comment]
**Benchmark mode kernel on Xeon Gold 5218**

- clang-9
```
                          benchmark         baseline        contender  change %
27         ModeKernelInt8/1048576/2  198.203 MiB/sec  245.465 MiB/sec    23.845  'null_percent': 50.0
26    ModeKernelInt16/1048576/10000  329.690 MiB/sec  351.171 MiB/sec     6.516  'null_percent': 0.01
9       ModeKernelInt16/1048576/100  323.647 MiB/sec  344.291 MiB/sec     6.379  'null_percent': 1.0
13        ModeKernelInt64/1048576/1  978.100 GiB/sec    1.016 TiB/sec     6.374 'null_percent': 100.0
1         ModeKernelInt64/1048576/0    1.671 GiB/sec    1.775 GiB/sec     6.253 , 'null_percent': 0.0
28        ModeKernelInt16/1048576/1  993.549 GiB/sec    1.023 TiB/sec     5.407 'null_percent': 100.0
24        ModeKernelInt32/1048576/0  928.620 MiB/sec  978.819 MiB/sec     5.406  'null_percent': 0.0
29        ModeKernelInt8/1048576/10  253.460 MiB/sec  267.073 MiB/sec     5.371  'null_percent': 10.0
19        ModeKernelInt32/1048576/2  425.023 MiB/sec  447.163 MiB/sec     5.209  'null_percent': 50.0
20        ModeKernelInt16/1048576/2  189.973 MiB/sec  199.458 MiB/sec     4.993  'null_percent': 50.0
5      ModeKernelBoolean/1048576/10   26.454 MiB/sec   27.770 MiB/sec     4.973 null_percent': 10.0
0          ModeKernelInt8/1048576/0  257.221 MiB/sec  269.917 MiB/sec     4.936 'null_percent': 0.0
12      ModeKernelInt32/1048576/100  719.901 MiB/sec  747.060 MiB/sec     3.773 'null_percent': 1.0
14         ModeKernelInt8/1048576/1  693.002 GiB/sec  717.353 GiB/sec     3.514 'null_percent': 100.0
17        ModeKernelInt16/1048576/0  484.109 MiB/sec  499.499 MiB/sec     3.179 'null_percent': 0.0
8        ModeKernelInt8/1048576/100  246.840 MiB/sec  253.669 MiB/sec     2.767 'null_percent': 1.0
23       ModeKernelInt32/1048576/10  654.627 MiB/sec  671.400 MiB/sec     2.562  'null_percent': 10.0
16      ModeKernelBoolean/1048576/2   27.659 MiB/sec   28.330 MiB/sec     2.427 null_percent': 50.0
25    ModeKernelInt32/1048576/10000  736.172 MiB/sec  753.046 MiB/sec     2.292  'null_percent': 0.01
2        ModeKernelInt16/1048576/10  305.731 MiB/sec  309.157 MiB/sec     1.121  'null_percent': 10.0
6         ModeKernelInt32/1048576/1    1.017 TiB/sec    1.018 TiB/sec     0.102 'null_percent': 100.0
7       ModeKernelBoolean/1048576/1  957.936 GiB/sec  956.161 GiB/sec    -0.185 'null_percent': 100.0
21    ModeKernelBoolean/1048576/100   26.571 MiB/sec   26.487 MiB/sec    -0.318  'null_percent': 1.0
10     ModeKernelInt8/1048576/10000  250.462 MiB/sec  245.625 MiB/sec    -1.931  'null_percent': 0.01
4       ModeKernelBoolean/1048576/0   28.303 MiB/sec   27.568 MiB/sec    -2.597  'null_percent': 0.0
22       ModeKernelInt64/1048576/10    1.265 GiB/sec    1.224 GiB/sec    -3.270  'null_percent': 10.0
18  ModeKernelBoolean/1048576/10000   26.428 MiB/sec   25.544 MiB/sec    -3.344 null_percent': 0.01
15    ModeKernelInt64/1048576/10000    1.385 GiB/sec    1.338 GiB/sec    -3.389  'null_percent': 0.01
3       ModeKernelInt64/1048576/100    1.408 GiB/sec    1.333 GiB/sec    -5.332 , 'null_percent': 1.0
11        ModeKernelInt64/1048576/2  861.695 MiB/sec  782.062 MiB/sec    -9.241  'null_percent': 50.0
```

- gcc-7.5
```
                          benchmark         baseline        contender  change %
5         ModeKernelInt32/1048576/2  361.224 MiB/sec  422.818 MiB/sec    17.052 'null_percent': 50.0
9         ModeKernelInt64/1048576/0    1.434 GiB/sec    1.650 GiB/sec    15.059  'null_percent': 0.0
25    ModeKernelInt32/1048576/10000  620.589 MiB/sec  712.089 MiB/sec    14.744 'null_percent': 0.01
28      ModeKernelInt32/1048576/100  621.477 MiB/sec  707.153 MiB/sec    13.786  'null_percent': 1.0
14       ModeKernelInt32/1048576/10  576.223 MiB/sec  646.869 MiB/sec    12.260 'null_percent': 10.0
21         ModeKernelInt8/1048576/2  176.230 MiB/sec  197.465 MiB/sec    12.049 'null_percent': 50.0
11        ModeKernelInt16/1048576/0  362.404 MiB/sec  405.761 MiB/sec    11.964  'null_percent': 0.0
22        ModeKernelInt64/1048576/1  958.410 GiB/sec    1.014 TiB/sec     8.374 null_percent': 100.0
1          ModeKernelInt8/1048576/1  719.935 GiB/sec  774.349 GiB/sec     7.558 null_percent': 100.0
26        ModeKernelInt32/1048576/1  984.700 GiB/sec    1.029 TiB/sec     7.053 null_percent': 100.0
29        ModeKernelInt64/1048576/2  668.946 MiB/sec  715.600 MiB/sec     6.974 'null_percent': 50.0
15         ModeKernelInt8/1048576/0  218.479 MiB/sec  232.068 MiB/sec     6.220  'null_percent': 0.0
2        ModeKernelInt64/1048576/10    1.113 GiB/sec    1.173 GiB/sec     5.421 'null_percent': 10.0
0     ModeKernelInt64/1048576/10000    1.285 GiB/sec    1.329 GiB/sec     3.427 'null_percent': 0.01
16    ModeKernelInt16/1048576/10000  344.989 MiB/sec  355.943 MiB/sec     3.175 'null_percent': 0.01
8       ModeKernelBoolean/1048576/1  965.557 GiB/sec  988.873 GiB/sec     2.415 null_percent': 100.0
20        ModeKernelInt8/1048576/10  233.980 MiB/sec  238.368 MiB/sec     1.875 'null_percent': 10.0
18      ModeKernelInt64/1048576/100    1.274 GiB/sec    1.292 GiB/sec     1.436  'null_percent': 1.0
23    ModeKernelBoolean/1048576/100   26.822 MiB/sec   27.067 MiB/sec     0.913  'null_percent': 1.0
24     ModeKernelInt8/1048576/10000  230.805 MiB/sec  232.875 MiB/sec     0.897 'null_percent': 0.01
17       ModeKernelInt16/1048576/10  321.025 MiB/sec  323.517 MiB/sec     0.776 'null_percent': 10.0
7       ModeKernelBoolean/1048576/0   27.920 MiB/sec   28.081 MiB/sec     0.578  'null_percent': 0.0
12  ModeKernelBoolean/1048576/10000   27.312 MiB/sec   27.314 MiB/sec     0.008 'null_percent': 0.01
13        ModeKernelInt16/1048576/1    1.012 TiB/sec    1.011 TiB/sec    -0.070 null_percent': 100.0
10      ModeKernelInt16/1048576/100  353.562 MiB/sec  353.029 MiB/sec    -0.151  'null_percent': 1.0
3        ModeKernelInt8/1048576/100  232.177 MiB/sec  231.227 MiB/sec    -0.409  'null_percent': 1.0
19        ModeKernelInt16/1048576/2  214.646 MiB/sec  211.749 MiB/sec    -1.350 'null_percent': 50.0
4      ModeKernelBoolean/1048576/10   28.997 MiB/sec   28.319 MiB/sec    -2.339 'null_percent': 10.0
6         ModeKernelInt32/1048576/0  803.848 MiB/sec  739.389 MiB/sec    -8.019  'null_percent': 0.0
27      ModeKernelBoolean/1048576/2   26.320 MiB/sec   23.780 MiB/sec    -9.648 'null_percent': 50.0
```

**Comment by cyb70289** [comment]
CI failure is about aws sdk: https://github.com/apache/arrow/pull/8232/checks?check_run_id=1142379049#step:8:3059

**Comment by cyb70289** [comment]
CI failure is about flight test. Looks not related.
https://ci.appveyor.com/project/ApacheSoftwareFoundation/arrow/builds/35331179/job/i7l5oi2mxnwytd4q#L1788

File: cpp/src/arrow/compute/kernel.h [hunk 1]
```diff
@@ -664,7 +664,7 @@ struct VectorKernel : public ArrayKernel {
 using ScalarAggregateConsume = std::function<void(KernelContext*, const ExecBatch&)>;
 
 using ScalarAggregateMerge =
-    std::function<void(KernelContext*, const KernelState&, KernelState*)>;
+    std::function<void(KernelContext*, KernelState&, KernelState*)>;
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernel.h:667]: If it can be moved, shouldn't you pass `KernelState&&` instead?

Comment (cyb70289) [reply cpp/src/arrow/compute/kernel.h:667]: Yes. Pass rvalue makes it more explicit. Will change.

Comment (cyb70289) [reply cpp/src/arrow/compute/kernel.h]: done

**Review by pitrou — APPROVED** [review]
+1, thank you

**Pull request closed without merging**